sstables/storage.cc

```diff
@@ -36,10 +36,57 @@
 
 namespace sstables {
 
+class dir_wrapper final {
+    std::filesystem::path _pathname;
+    file _file;
+    bool _file_stale;
+
+public:
+    explicit dir_wrapper(std::filesystem::path pathname) : _pathname(std::move(pathname)) {};
+    dir_wrapper(const dir_wrapper&) = delete;
+    dir_wrapper& operator=(const dir_wrapper&) = delete;
+    dir_wrapper(dir_wrapper&&) = delete;
+    dir_wrapper& operator=(dir_wrapper&&) = delete;
+    ~dir_wrapper() = default;
+
+    const std::filesystem::path::string_type& native() const noexcept {
+        return _pathname.native();
+    }
+
+    operator std::filesystem::path() const noexcept {
+        return _pathname;
+    }
+
+    dir_wrapper& operator=(const sstring &nd) {
+        _pathname = nd;
+        _file_stale = true;
+        return *this;
+    }
+
```

Comment:
> `future<> change_dir(new_dir) {`

... also I meant to preserve the prior assignment syntax (`_dir = ...`).

Reply:
I'd consider keeping a future member for close that could be use to closing the file in the background after such assignment.

So if `close()` would look like this:
```
future<> close() noexcept {
	if (_file) {
            _close_fut = _close_fut.then([f = _file.close()] () mutable {
                return std::move(f);
            }
        }
        return std::exchange(_close_fut, make_ready_future());
}
```

and then this synchronous assignment operator could be implemented like:
```
    dir_wrapper& operator=(const sstring &nd) {
        (void)close();
        _pathname = nd;
        return *this;
    }
```

and instead of `_file_stale` you can just probe the file using its boolean operator to see if it's opened or not.

Reply:
> I'd consider keeping a future member for close that could be use to closing the file in the background after such assignment.
> 
> So if `close()` would look like this:
> 
> ```
> future<> close() noexcept {
> 	if (_file) {
>             _close_fut = _close_fut.then([f = _file.close()] () mutable {
>                 return std::move(f);
>             }
>         }
>         return std::exchange(_close_fut, make_ready_future());
> }
> ```
> 
> and then this synchronous assignment operator could be implemented like:
> 
> ```
>     dir_wrapper& operator=(const sstring &nd) {
>         (void)close();
>         _pathname = nd;
>         return *this;
>     }
> ```
> 
> and instead of `_file_stale` you can just probe the file using its boolean operator to see if it's opened or not.

Can you please walk me through this in detail? AFAICT, `_close_fut` would be initialized as a ready future (?). Upon `close()`, we'd chain the actual close operation to this (originally ready) future (?), then effectively throw away `_close_fut` (nobody would ever wait for the close to resolve). This looks complicated to me (and somewhat counter to `future` being marked `[[nodiscard]]` in "seastar/include/seastar/core/future.hh" -- I guess that's why the `(void)` cast is required too).

Furthermore, I don't understand why a permanent `_close_fut` member is needed for this.

Reply:
> Can you please walk me through this in detail? AFAICT, `_close_fut` would be initialized as a ready future (?).

Right

> Upon `close()`, we'd chain the actual close operation to this (originally ready) future (?), then effectively throw away `_close_fut` (nobody would ever wait for the close to resolve).

Why throw it away?  close() should return it, and the caller of close should await it.

> This looks complicated to me (and somewhat counter to `future` being marked `[[nodiscard]]` in "seastar/include/seastar/core/future.hh" -- I guess that's why the `(void)` cast is required too).

Yes, the void cast is for dropping the future.
But it seemw like I was wrong to propose it.
The intention was to chain _file.close() to _close_fut, if needed (_file evaluates to true), so that the final close() call would pick it up and return it.

> Furthermore, I don't understand why a permanent `_close_fut` member is needed for this.

The _close_fut allows to chain file::close calls and wait for them on the final close() call.
this assignment: `_close_fut = _close_fut.then([f = std::move(_file)] () mutable { retunr f.close(); }` would consume _close_fut is it's available and synchronously return the future returned by the lambda arg to .then().

Reply:
@bhalevy I think I've addressed all other v4 comments, but I'm stuck on this one. I've slighty rewritten your code in order to explain it to myself:

~~~
    future<> close() noexcept {
	if (_file) {
            auto f = _file.close();
        
            _close_fut = _close_fut.then([f] () mutable {
                return std::move(f);
            }
        }
        return std::exchange(_close_fut, make_ready_future<>());
    }
~~~

There are several things I don't understand about this:

1. I don't know how to hook an `io_error_handler` into this, for `do_io_check()`.
2. I don't know who would call this member function (even once), now that the assignment-from-sstring operator is gone. (When `_dir` is reassigned from an sstring directory name, a new constructor takes effect now, which prepares a new `opened_directory` object for replacing `_dir`, and then the replacement occurs through the now-enabled move assignment operator.) This means that a `_close_fut` member of the `opened_directory` object that's being destructed would also be destructed, but that doesn't appear related to this new `close()` member function in any way.
3. I can't envisage under what circumstances *multiple calls* to `close` could be made. Again, currently there are zero call sites.
4. Provided there were multiple calls to `close()` in succession, I still wouldn't understand how the function was supposed to work:
    1. Assume `_file` evals to true upon entry to the first `close()` call.
    2. At that time, `_close_fut` is the empty ready future.
    3. We initiate `_file.close()` asynchronously, and save (capture) the future for it.
    4. We chain the resolution of this pending `_file.close()` after `_close_fut`.
    5. We return this new chain to the caller, and at the same time, we replace `_close_fut` with the empty ready future -- precisely what it was, upon entry to the function.
    6. Assume the second call to `close()` occurs now.
    7. If `_file` got closed meanwhile, then we do effectively nothing; we just replace the empty ready future in `_close_fut` with another instance of the empty ready future.
    8. Conversely, if `_file` is still open, then this second invocation behaves exactly the same way as the very first invocation; note that upon *second entry* to `close()`, `_close_fut` is invariably the empty ready future.
5. No matter at which point the destructor for `opened_directory` ran, it would always find the empty ready future in `_close_fut`.

I don't understand how the above process implements a "final close". The effect seems to be exactly the same as if we simply did

~~~
    future<> close() noexcept {
        return _file ? _file.close() : make_ready_future<>();
    }
~~~

... I feel tempted to submit v5 at this point, so that we have a more concrete basis for discussing this further. I've addressed all the other feedback. However, as I note [elsewhere](https://github.com/scylladb/scylladb/pull/19624#discussion_r1681625798), in v5 there is a small step-back relative to v4: namely it does not close the previous open directory in `co_await` context; it closes the prev dir synchronously, as a part of the (default) destructor logic.

What do you recommend, should I post v5, or should we continue discussing this one remaining point before that? Thanks!

Reply:
> What do you recommend, should I post v5, or should we continue discussing this one remaining point before that? Thanks!

It's hard to refer to things that are in the yet-to-be-published version, so better submit it sooner rather than later.

Generally speaking file.close() is asynchronous and we take care to have ab async close method to higher-level classes that hold a file member, like input and output streams.  The synchronous close syscall for the posix file is complimentary in nature.

That said, those async close functions are there for the purpose of waiting on asynchrouns background i/o (read-ahead or write-behinds), to prevent use-aftre-free, BUT, in this case the file object for the directory is not involved in async i/o so the synchronous close is less of a problem.

Let's start with a simple version that has no async close() function.  It can be added as a follow up as well.

Reply:
OK, I've pushed v5 now.

Meanwhile, I've started work on the async close; I just remembered @raphaelsc's [comment](https://github.com/scylladb/scylladb/pull/19624#pullrequestreview-2173477444).

Reply:
> Meanwhile, I've started work on the async close; I just remembered @raphaelsc's [comment](https://github.com/scylladb/scylladb/pull/19624#pullrequestreview-2173477444).

Should be implemented in v6 (which is a fast-forwardable descendant of v5).

Reply:
To keep the conversation manageable, I'm resolving this thread now. The changes I've done for it are listed in the v5 and v6 changelogs, so they won't be missed.